plugins/inputs/linux_cpu/linux_cpu.go

```diff
@@ -0,0 +1,211 @@
+
+  ## CPU metrics collected by the plugin.
+  ## Supported options:
+  ## "cpufreq", "thermal"
+  ## Defaults:
+  # metrics = ["cpufreq""]
+`
+
+func (g *LinuxCPU) SampleConfig() string {
+	return sampleConfig
+}
+
+func (g *LinuxCPU) Description() string {
+	return "Collects CPU metrics exposed on Linux"
+}
+
+func (g *LinuxCPU) Init() error {
+	if g.PathSysfs == "" {
+		g.PathSysfs = defaultHostSys
+	}
+
+	if len(g.Metrics) == 0 {
+		// The user has not enabled any of the metrics: fail fast
+		return fmt.Errorf("no metrics selected")
+	}
+
+	cpus, err := g.discoverCpus()
+	if err != nil {
+		return err
+	} else if len(cpus) == 0 {
+		// Although the user has specified metrics to collect, `discoverCpus` failed to find the required metrics
+		return fmt.Errorf("no CPUs detected to track")
+	}
+	g.cpus = cpus
+
+	return nil
+}
+
+func (g *LinuxCPU) Gather(acc telegraf.Accumulator) error {
+	for _, cpu := range g.cpus {
+		fields := make(map[string]interface{})
+		tags := map[string]string{"cpu": cpu.id}
+
+		failed := false
+		for name, fd := range cpu.props {
+			v, err := readUintFromFile(fd)
+			if err != nil {
+				acc.AddError(err)
+				failed = true
+				break
+			}
+
+			fields[name] = v
+		}
+
+		if !failed {
+			acc.AddFields("linux_cpu", fields, tags)
+		}
+	}
+
+	return nil
+}
+
+func (g *LinuxCPU) discoverCpus() ([]cpu, error) {
+	var cpus []cpu
+
+	glob := path.Join(g.PathSysfs, "devices/system/cpu/cpu[0-9]*")
+	cpuDirs, err := filepath.Glob(glob)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(cpuDirs) == 0 {
+		return nil, fmt.Errorf("no CPUs detected at: %s", glob)
+	}
+
+	for _, dir := range cpuDirs {
+		_, cpuName := filepath.Split(dir)
+		cpuNum := strings.TrimPrefix(cpuName, "cpu")
+
+		cpu := cpu{
+			id:    cpuNum,
+			path:  dir,
+			props: make(map[string]*os.File),
+		}
+
+		var props []prop
+
+		if contains(g.Metrics, cpufreq) {
+			props = append(props,
+				prop{ name: "scaling_cur_freq", path: "cpufreq/scaling_cur_freq", optional: false },
+				prop{ name: "scaling_min_freq", path: "cpufreq/scaling_min_freq", optional: false },
+				prop{ name: "scaling_max_freq", path: "cpufreq/scaling_max_freq", optional: false },
+				prop{ name: "cpuinfo_cur_freq", path: "cpufreq/cpuinfo_cur_freq", optional: true },
+				prop{ name: "cpuinfo_min_freq", path: "cpufreq/cpuinfo_min_freq", optional: true },
+				prop{ name: "cpuinfo_max_freq", path: "cpufreq/cpuinfo_max_freq", optional: true },
+			)
+		}
+
+		if contains(g.Metrics, thermal) {
+			props = append(
+				props,
+				prop{ name: "throttle_count", path: "thermal_throttle/core_throttle_count", optional: false },
+				prop{ name: "throttle_max_time", path: "thermal_throttle/core_throttle_max_time_ms", optional: false },
+				prop{ name: "throttle_total_time", path: "thermal_throttle/core_throttle_total_time_ms", optional: false },
+			)
+		}
+
+		var failed = false
+		for _, prop := range props {
+			fd, err := os.Open(filepath.Join(dir, prop.path))
```

Comment:
You open file here and you never close it...
Better idea would be to check here if file exists and is a regular file (check `validateDbPath` function in `ras.go` for example) and to save fullPath to file in `cpu.props`. Then in `readUintFromFile` you would open, read, close every file.

Reply:
Correct. I am a bit hesitant on having the collector perform multiple syscalls (for each of these metrics) every collection. With the current approach, a file descriptor is openend for each of the metrics and performs just one read call every collection cycle. The downside of this is that the process has more open file descriptors (possibly too many).